### PR TITLE
[consider-using-augmented-assign] Do not warn for non-commutative operators

### LINKED ISSUE #7639: Do not suggest augmented assigns for non-commutative operations with the target as the right operand
### Bug description

Pylint suggests augmented assigns for non-commutative binary operations, even when the assignment target is the right operand.

~~~Python
"""Minimal example where a C0210 message is displayed even though a simple
transformation does not seem possible.
"""

# pylint: disable=invalid-name

x = 1
x = 1 - x
x = 2 / x
~~~

### Configuration

_No response_

### Command used

```shell
pylint c0210_issue.py
```

### Pylint output

```shell
************* Module c0210_issue
c0210_issue.py:8:0: C0210: Use '-=' to do an augmented assign directly (consider-using-augmented-assign)
c0210_issue.py:9:0: C0210: Use '/=' to do an augmented assign directly (consider-using-augmented-assign)

------------------------------------------------------------------
Your code has been rated at 3.33/10 (previous run: 3.33/10, +0.00)
```

### Expected behavior

No error should be displayed.

### Pylint version

```shell
pylint 2.16.0-dev
astroid 2.12.10
Python 3.8.5 (default, Sep  7 2020, 16:44:59) [GCC 6.3.0 20170516]
```

### OS / Environment

Debian 9.13 (stretch)

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Or rather, only warn for known commutative operators.

Closes #7639

### CODE REVIEW

**Review by DanielNoord — DISMISSED** [review]
Love it!

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -2120,7 +2121,11 @@ def is_augmented_assign(node: nodes.Assign) -> tuple[bool, str]:
 
     if _is_target_name_in_binop_side(target, binop.left):
         return True, binop.op
-    if _is_target_name_in_binop_side(target, binop.right):
+    if (
+        _is_target_name_in_binop_side(target, binop.right)
+        # Unless an operator is commutative, we should not raise (i.e. x = 3/x)
+        and binop.op in COMMUTATIVE_OPERATORS
```
Comment (nickdrozd) [inline pylint/checkers/utils.py:2127]:
```suggestion
        # Unless an operator is commutative, we should not raise (i.e. x = 3/x)
        binop.op in COMMUTATIVE_OPERATORS
        and _is_target_name_in_binop_side(target, binop.right)
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 2]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
 my_str = Unknown % my_str
 my_str = my_str % Unknown  # [consider-using-augmented-assign]
 
+# subtraction is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:67]:
```suggestion
# subtraction is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 3]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x - 3  # [consider-using-augmented-assign]
+x = 3 - x
+
+# multiplication is commutative
 x = x * 3  # [consider-using-augmented-assign]
+x = 3 * x  # [consider-using-augmented-assign]
+
+# division is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:75]:
```suggestion
# division is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 4]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x / 3  # [consider-using-augmented-assign]
+x = 3 / x
+
+# integer division is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:79]:
```suggestion
# integer division is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 5]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x // 3  # [consider-using-augmented-assign]
+x = 3 // x
+
+# Left shift operator is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:83]:
```suggestion
# Left shift operator is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 6]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x << 3  # [consider-using-augmented-assign]
+x = 3 << x
+
+# Right shift operator is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:87]:
```suggestion
# Right shift operator is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 7]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x >> 3  # [consider-using-augmented-assign]
+x = 3 >> x
+
+# modulo is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:91]:
```suggestion
# modulo is not commutative
```

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py [hunk 8]
```diff
@@ -61,17 +64,58 @@ def return_str() -> str:
...
 x = x % 3  # [consider-using-augmented-assign]
+x = 3 % x
+
+# exponential is anti-commutative
```
Comment (mbyrnepr2) [inline tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:95]:
```suggestion
# exponential is not commutative
```

**Review by mbyrnepr2 — COMMENTED** [review]
Took a quick look. I think the `anti-commutative` phrase may be misleading/mean something else (https://www.dictionary.com/browse/anticommutative#:~:text=adjective%20Mathematics.,%2C%20as%20ab%20%3D%20%E2%88%92ba.)

File: tests/functional/ext/code_style/cs_consider_using_augmented_assign.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply tests/functional/ext/code_style/cs_consider_using_augmented_assign.py:67]: This one is really anti-commutative, this is the one I extrapoled from :)

**Review by Pierre-Sassoulas — COMMENTED** [review]
Nice catch @mbyrnepr2 !

**Review by mbyrnepr2 — APPROVED** [review]

**Pull request merged**